### ISSUE TITLE
Rule Change: Treat class / instance fields as named constants in no-magic-numbers

### ISSUE THREAD

**Issue by terrablue** [eslint/eslint#16228]
### What rule do you want to change?

no-magic-numbers

### What change to do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
class Foo {
  static bar = 1;
  baz = 2;
}
```

### What does the rule currently do for this code?

Shows these two errors:
`No magic number: 1`
`No magic number: 2`

### What will the rule do after it's changed?

Show no errors.

### Participation

- [X] I am willing to submit a pull request to implement this change.

**Comment by mdjermanovic** [comment]
Hi @terrablue, thanks for the issue!

This rule has an option [`detectObjects`](https://eslint.org/docs/latest/rules/no-magic-numbers#detectobjects):

> A boolean to specify if we should detect numbers when setting object properties

Default is `false`:

```js
/* eslint no-magic-numbers: "error" */

class Foo {
  constructor() {
    this.baz = 2; // ok
  }
}

Foo.bar = 1; // ok

const obj = {
  foo: 3 // ok
}

```

I think the `detectObjects` option should also apply to class field initializers (perhaps we missed considering this rule in https://github.com/eslint/eslint/issues/14857), but would like more opinions on this.

**Comment by terrablue** [comment]
I agree, since it's declarative in nature and represents an initial value, like `let baz = 2;`, I think it should be treated the same.

**Comment by nzakas** [comment]
I think the utility of this rule is questionable at best so I’d rather not make any further changes to it.

**Comment by ljharb** [comment]
Many of us find the utility of this rule quite high.

**Comment by ljharb** [comment]
bump, in hopes this useful rule can be improved.

**Comment by nzakas** [comment]
@eslint/eslint-tsc does anyone want to champion this change? If not, we should close this.

**Comment by mdjermanovic** [comment]
I'll champion this.

I think the best solution is to add a new option: `ignoreClassFieldInitialValues` (default `false` to keep the current default behavior).

**Comment by nzakas** [comment]
All right, marking as accepted.

**Comment by mdjermanovic** [comment]
I'm working on this.
